README.md
# TweetWebView for Flutter 

[![Build Status](https://travis-ci.org/larryaasen/tweet_webview.svg?branch=master)](https://travis-ci.org/larryaasen/tweet_webview)    [![codecov](https://codecov.io/gh/larryaasen/tweet_webview/branch/master/graph/badge.svg)](https://codecov.io/gh/larryaasen/tweet_webview) [![pub package](https://img.shields.io/pub/v/tweet_webview.svg)](https://pub.dartlang.org/packages/tweet_webview)

TweetWebView is a Flutter widget to easily display a Twitter tweet in a WebView widget using [webview_flutter](https://pub.dartlang.org/packages/webview_flutter).

## Example 1
Display a single tweet in a TweetWebView widget using the tweet URL.

```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 1'),
      ),
      body: TweetWebView.tweetUrl("https://twitter.com/Interior/status/463440424141459456"),
    );
  }
}
```

[<img src="screenshots/example1.png" width="300" alt="Screenshot of list of tweets" style="border:1px solid black">](screenshots/example1.png)

## Example 2
Display a single tweet in a TweetWebView widget using the tweet ID.

```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 2'),
      ),
      body: TweetWebView.tweetID('463440424141459456'),
    );
  }
}
```

[<img src="screenshots/example2.png" width="300" alt="Screenshot of list of tweets" style="border:1px solid black">](screenshots/example2.png)

## Example 3
Display a list of tweets in TweetWebView widgets using the tweet IDs.

```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 3'),
      ),
      body: _buildBody(),
    );
  }

  Widget _buildBody() {
    final tweets = ['1065424382292566017', '1068219397293125633', '1068551446029832192', '1065362290512293888'];

    final list = ListView.builder(
      scrollDirection: Axis.vertical,
      padding: EdgeInsets.fromLTRB(0.0, 0.0, 0.0, 0.0),
      itemCount: tweets.length,
      itemBuilder: (context, index) {
        var tweetID = tweets[index];
        return Card(
          child: TweetWebView.tweetID(tweetID),
        );
      },
    );

    final container = Container(
        color: Colors.black26,
        child: Center(child: list)
    );

    return container;

  }
}
```

[<img src="screenshots/example3.png" width="300" alt="Screenshot of list of tweets">](screenshots/example3.png)

## Contributing
All [comments](https://github.com/larryaasen/tweet_webview/issues) and [pull requests](https://github.com/larryaasen/tweet_webview/pulls) are welcome.

CHANGELOG.md
## 0.2.2

* Updated pubspec to support the latest dependencies, mainly path_provider.

## 0.2.1

* Updated dependency of webview_flutter package to ^0.3.3+1

## 0.2.0

* Updated dependency of webview_flutter package to 0.2.0

## 0.1.0

* Initial Open Source release.


test/tweet_webview_test.dart
/*
 * Copyright (c) 2018 Larry Aasen. All rights reserved.
 */

import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:tweet_webview/tweet_webview.dart';

void main() {
  testWidgets('Widget test', (WidgetTester tester) async {
    await tester.pumpWidget(MyWidget());

    expect(find.text('TweetWebView test'), findsOneWidget);
    expect(find.text('Loading...'), findsNWidgets(4));

    sleep(const Duration(seconds: 1));

    expect(find.text('Loading...'), findsNWidgets(4));
  });
}

class MyWidget extends StatelessWidget {
  const MyWidget({
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final tweet1 = TweetWebView(
        tweetUrl: 'https://twitter.com/Interior/status/463440424141459456');
    final tweet2 = TweetWebView(tweetID: '463440424141459456');
    final tweet3 = TweetWebView.tweetUrl(
        'https://twitter.com/Interior/status/463440424141459456');
    final tweet4 = TweetWebView.tweetID('463440424141459456');
    return MaterialApp(
      title: 'TweetWebView test',
      home: Scaffold(
        appBar: AppBar(
          title: Text('TweetWebView test'),
        ),
        body: Column(
          children: <Widget>[tweet1, tweet2, tweet3, tweet4],
        ),
      ),
    );
  }
}


example/example.dart
/*
 * Copyright (c) 2018 Larry Aasen. All rights reserved.
 */

import 'package:flutter/material.dart';
import 'package:tweet_webview/tweet_webview.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Tweet WebView Example 1',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 1'),
      ),
      body: TweetWebView.tweetUrl(
          "https://twitter.com/Interior/status/1172258799576653825"),
    );
  }
}


example/example2.dart
/*
 * Copyright (c) 2018 Larry Aasen. All rights reserved.
 */

import 'package:flutter/material.dart';
import 'package:tweet_webview/tweet_webview.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Tweet WebView Example 2',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 2'),
      ),
      body: TweetWebView.tweetID('463440424141459456'),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:tweet_webview/tweet_webview.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Tweet WebView Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Tweet WebView Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        color: Colors.black12,
        child:Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Container(
                height: 400.0,
                  margin: EdgeInsets.all(20.0),
                  child: TweetWebView(tweetUrl: 'https://twitter.com/Interior/status/463440424141459456'))
            ],
          ),
        ),
      ),
    );
  }
}


lib/tweet_webview.dart
/*
 * Copyright (c) 2018 Larry Aasen. All rights reserved.
 */

library tweet_webview;

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:flutter/material.dart';
import 'package:path_provider/path_provider.dart';
import 'package:webview_flutter/webview_flutter.dart';

class TweetWebView extends StatefulWidget {
  final String tweetUrl;

  final String tweetID;

  TweetWebView({this.tweetUrl, this.tweetID});

  TweetWebView.tweetID(String tweetID)
      : this.tweetID = tweetID,
        this.tweetUrl = null;
  TweetWebView.tweetUrl(String tweetUrl)
      : this.tweetUrl = tweetUrl,
        this.tweetID = null;

  @override
  _TweetWebViewState createState() => new _TweetWebViewState();
}

class _TweetWebViewState extends State<TweetWebView> {
  String _tweetHTML;
  String _filename;

  @override
  void initState() {
    super.initState();

    _requestTweet();
  }

  @override
  Widget build(BuildContext context) {
    var child;
    if (_tweetHTML != null && _tweetHTML.length > 0) {
      final downloadUrl = Uri.file(_filename).toString();

      // Create the WebView to contian the tweet HTML
      final webView = WebView(
          initialUrl: downloadUrl, javascriptMode: JavascriptMode.unrestricted);

      // The WebView creates an exception: RenderAndroidView object was given an infinite size during layout.
      // To avoid that exception a max height constraint will be used. Hopefully soon the WebView will be able
      // to size itself so it will not have an infinite height.
      final box = LimitedBox(
        maxHeight: 500.0,
        child: webView,
      );

      child = box;
    } else {
      child = Text('Loading...');
    }

    return Container(child: child);
  }

  /// Download the embedded tweet.
  /// See Twitter docs: https://developer.twitter.com/en/docs/twitter-for-websites/embedded-tweets/overview
  void _requestTweet() async {
    String tweetUrl = widget.tweetUrl;
    String tweetID;

    if (tweetUrl == null || tweetUrl.isEmpty) {
      if (widget.tweetID == null || widget.tweetID.isEmpty) {
        throw new ArgumentError('Missing tweetUrl or tweetID property.');
      }
      tweetUrl = _formTweetURL(widget.tweetID);
      tweetID = widget.tweetID;
    }

    if (tweetID == null) {
      tweetID = _tweetIDFromUrl(tweetUrl);
    }

    // Example: https://publish.twitter.com/oembed?url=https://twitter.com/Interior/status/463440424141459456
    final downloadUrl = "https://publish.twitter.com/oembed?url=$tweetUrl";
    print("TweetWebView._requestTweet: $downloadUrl");

    final jsonString = await _loadTweet(downloadUrl);
    final html = _parseTweet(jsonString);
    if (html != null) {
      final filename = await _saveTweetToFile(tweetID, html);
      setState(() {
        _tweetHTML = html;
        _filename = filename;
      });
    }
  }

  String _tweetIDFromUrl(String tweetUrl) {
    final uri = Uri.parse(tweetUrl);
    if (uri.pathSegments.length > 0) {
      return uri.pathSegments[uri.pathSegments.length - 1];
    }
    return null;
  }

  String _formTweetURL(String tweetID) {
    return "https://twitter.com/Interior/status/$tweetID";
  }

  Future<String> _saveTweetToFile(String tweetID, String html) async {
    Directory tempDir = await getTemporaryDirectory();
    String tempPath = tempDir.path;
    final filename = '$tempPath/tweet-$tweetID.html';
    File(filename).writeAsString(html);
    return filename;
  }

  String _parseTweet(String jsonString) {
    if (jsonString == null || jsonString.isEmpty) {
      print('TweetWebView._parseTweet: empty jsonString');
      return null;
    }

    var item;
    try {
      item = json.decode(jsonString);
    } catch (e) {
      print(e);
      print('error parsing tweet json: $jsonString');
      return '<p>error loading tweet</p>';
    }

    final String html = item['html'];

    if (html == null || html.isEmpty) {
      print('TweetWebView._parseTweet: empty html');
    }

    return html;
  }

  Future<String> _loadTweet(String tweetUrl) async {
    http.Response result = await _downloadTweet(tweetUrl);

    return result.body;
  }

  Future<http.Response> _downloadTweet(String tweetUrl) {
    return http.get(tweetUrl);
  }
}


example/lib/example1.dart
/*
 * Copyright (c) 2018 Larry Aasen. All rights reserved.
 */

import 'package:flutter/material.dart';
import 'package:tweet_webview/tweet_webview.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Tweet WebView Example 1',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 1'),
      ),
      body: TweetWebView.tweetUrl("https://twitter.com/Interior/status/463440424141459456"),
    );
  }
}


example/lib/example3.dart
/*
 * Copyright (c) 2018 Larry Aasen. All rights reserved.
 */

import 'package:flutter/material.dart';
import 'package:tweet_webview/tweet_webview.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Tweet WebView Example 3',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tweet WebView Example 3'),
      ),
      body: _buildBody(),
    );
  }

  Widget _buildBody() {
    final tweets = [
      '1065424382292566017',
      '1068219397293125633',
      '1068551446029832192',
      '1065362290512293888'
    ];

    final list = ListView.builder(
      scrollDirection: Axis.vertical,
      padding: EdgeInsets.fromLTRB(0.0, 0.0, 0.0, 0.0),
      itemCount: tweets.length,
      itemBuilder: (context, index) {
        var tweetID = tweets[index];
        return Card(
          child: TweetWebView.tweetID(tweetID),
        );
      },
    );

    final container =
        Container(color: Colors.black26, child: Center(child: list));

    return container;
  }
}
